File: programmers/kakao/후보키.java
import java.util.*;

class Solution {
    public int solution(String[][] relation) {
int n = relation.length;
		int m = relation[0].length;
		ArrayList<Integer> ans = new ArrayList<Integer>();

		for (int i = 1; i < 1 << m; i++) {
			Set<String> set = new HashSet<String>();

			for (int j = 0; j < n; j++) {
				String sb = new String();
				for (int k = 0; k < m; k++) {
					if ((i & (1 << k)) > 0) {
						sb += relation[j][k];
					}
				}
				set.add(sb);
			}
			if (set.size() == n && possible(ans, i)) {
				ans.add(i);
			}
		}
        
        
		return ans.size();
	}

	public static boolean possible(ArrayList<Integer> ans, int x) {
		for (int i = 0; i < ans.size(); i++) {
			if ((ans.get(i) & x) == ans.get(i)) {
				return false;

			}
		}
		return true;
	}
}

File: programmers/kakao/후보키.js
function solution(arr) {
  let len = arr.length;
  let count = 0;
  let ans = [];

  for (let i = 1; i < 1 << arr[0].length; i++) {
    let set = new Set();
    for (let j = 0; j < len; j++) {
      let string = '';
      for (let k = 0; k < arr[0].length; k++) {
        if (i & (1 << k)) {
          string += arr[j][k];
        }
      }
      set.add(string);
    }
    if (set.size === len && possible(ans, i)) {
      ans.push(i);
    }
  }

  console.log(ans.length);
}

function possible(ans, x) {
  for (let i = 0; i < ans.length; i++) {
    if ((ans[i] & x) === ans[i]) {
      return false;
    }
  }
  return true;
}

solution([
  ['100', 'ryan', 'music', '2'],
  ['200', 'apeach', 'math', '2'],
  ['300', 'tube', 'computer', '3'],
  ['400', 'con', 'computer', '4'],
  ['500', 'muzi', 'music', '3'],
  ['600', 'apeach', 'music', '2']
]);
